packages/camera/camera/example/ios/RunnerTests/CameraMethodChannelTests.m

```diff
@@ -0,0 +1,101 @@
+#import <OCMock/OCMock.h>
+
+@interface FLTThreadSafeFlutterResult ()
+@property(readonly, nonatomic) FlutterResult flutterResult;
+@end
+
+@interface MockFLTThreadSafeFlutterResult : FLTThreadSafeFlutterResult
+@property(readonly, nonatomic) NSNotificationCenter *notificationCenter;
+@property(nonatomic, nullable) id receivedResult;
+@end
+
+@implementation MockFLTThreadSafeFlutterResult
+/**
+ Initialize with a notification center.
+ */
+- (id)initWithNotificationCenter:(NSNotificationCenter *)notificationCenter {
+  self = [super init];
+  _notificationCenter = notificationCenter;
+  return self;
+}
+
+/**
+ Called when result is successful. Sends "successWithData" to the notification center.
+ */
+- (void)successWithData:(id)data {
+  _receivedResult = data;
+  [self->_notificationCenter postNotificationName:@"successWithData" object:nil];
+}
+@end
+
+@interface CameraPlugin (Test)
+- (void)handleMethodCallWithThreadSafeResult:(FlutterMethodCall *)call
+                                      result:(FLTThreadSafeFlutterResult *)result;
+@end
+
+@interface CameraMethodChannelTests : XCTestCase
+@property(readonly, nonatomic) CameraPlugin *camera;
+@property(readonly, nonatomic) MockFLTThreadSafeFlutterResult *resultObject;
+@property(readonly, nonatomic) NSNotificationCenter *notificationCenter;
+@end
+
+@implementation CameraMethodChannelTests
+
+- (void)setUp {
+  _camera = [[CameraPlugin alloc] init];
+  _notificationCenter = [[NSNotificationCenter alloc] init];
+
+  // Setup mocks for initWithCameraName method
+  id avCaptureDeviceInputMock = OCMClassMock([AVCaptureDeviceInput class]);
+  OCMStub([avCaptureDeviceInputMock deviceInputWithDevice:[OCMArg any] error:[OCMArg anyObjectRef]])
+      .andReturn([AVCaptureInput alloc]);
+
+  id avCaptureSessionMock = OCMClassMock([AVCaptureSession class]);
+  OCMStub([avCaptureSessionMock alloc]).andReturn(avCaptureSessionMock);
+  OCMStub([avCaptureSessionMock canSetSessionPreset:[OCMArg any]]).andReturn(YES);
+
+  _resultObject =
+      [[MockFLTThreadSafeFlutterResult alloc] initWithNotificationCenter:_notificationCenter];
+}
+
+- (void)tearDown {
+  // Put teardown code here. This method is called after the invocation of each test method in the
+  // class.
+}
+
+- (void)testCreate_ShouldCallResultOnMainThread {
+  // Setup method call
+  XCTNSNotificationExpectation *notificationExpectation =
+      [[XCTNSNotificationExpectation alloc] initWithName:@"successWithData"
+                                                  object:nil
+                                      notificationCenter:_notificationCenter];
+
+  FlutterMethodCall *call = [FlutterMethodCall
+      methodCallWithMethodName:@"create"
+                     arguments:@{@"resolutionPreset" : @"medium", @"enableAudio" : @(1)}];
+
+  // Call handleMethodCall
+  [_camera handleMethodCallWithThreadSafeResult:call result:_resultObject];
+
+  // Don't expect a result yet
+  XCTAssertNil(_resultObject.receivedResult);
```

Comment:
It's not clear to me why this is an assertion. Why would it be an error if the plugin were to respond synchronously?

Reply:
The create call does some heavy work which should not be on the UI thread, so if this responds synchronously that means the calls on the background thread are now done on the main thread, or that the function completes before the calls on the background thread are completed.

Reply:
What if it's done on the background thread but the background thread work completes before the main thread is scheduled again to run this assert?

Reply:
Didn't expect that could happen, doesn't this run synchronously? If that can happen I'll remove the assert.

Reply:
Hm, I guess even if the background thread work completes, the part that was kicked back to the main thread won't have been run yet. So it's fine to leave, as long as you add a comment explaining why you're asserting that it *can't* have finished, since that's unusual (usually callbacks can happen at any time, including synchronously).